lib/carbon/writer.py

```diff
@@ -42,121 +54,188 @@
-
-        if not archiveConfig:
-          raise Exception("No storage schema matched the metric '%s', check your storage-schemas.conf file." % metric)
-
-        if settings.LOG_CREATES:
-          log.creates("creating database metric %s (archive=%s xff=%s agg=%s)" %
-                      (metric, archiveConfig, xFilesFactor, aggregationMethod))
-        try:
-            state.database.create(metric, archiveConfig, xFilesFactor, aggregationMethod)
-            instrumentation.increment('creates')
-        except Exception as e:
-            log.err()
-            log.msg("Error creating %s: %s" % (metric, e))
-            instrumentation.increment('errors')
-            continue
-      # If we've got a rate limit configured lets makes sure we enforce it
-      if UPDATE_BUCKET:
-        UPDATE_BUCKET.drain(1, blocking=True)
+    (metric, datapoints) = cache.drain_metric()
+    if metric is None:
+      # end the loop
+      break
+
+    dbFileExists = state.database.exists(metric)
+
+    if not dbFileExists:
+      if CREATE_BUCKET and not CREATE_BUCKET.drain(1):
+        # If our tokenbucket doesn't have enough tokens available to create a new metric
+        # file then we'll just drop the metric on the ground and move on to the next
+        # metric.
+        # XXX This behavior should probably be configurable to no tdrop metrics
+        # when rate limitng unless our cache is too big or some other legit
+        # reason.
+        continue
+
+      archiveConfig = None
+      xFilesFactor, aggregationMethod = None, None
+
+      for schema in SCHEMAS:
+        if schema.matches(metric):
+          if settings.LOG_CREATES:
+            log.creates('new metric %s matched schema %s' % (metric, schema.name))
+          archiveConfig = [archive.getTuple() for archive in schema.archives]
+          break
+
+      for schema in AGGREGATION_SCHEMAS:
+        if schema.matches(metric):
+          if settings.LOG_CREATES:
+            log.creates('new metric %s matched aggregation schema %s'
+                        % (metric, schema.name))
+          xFilesFactor, aggregationMethod = schema.archives
+          break
+
+      if not archiveConfig:
+        raise Exception("No storage schema matched the metric '%s', check your storage-schemas.conf file." % metric)
+
+      if settings.LOG_CREATES:
+        log.creates("creating database metric %s (archive=%s xff=%s agg=%s)" %
+                    (metric, archiveConfig, xFilesFactor, aggregationMethod))
       try:
-        t1 = time.time()
-        # If we have duplicated points, always pick the last. update_many()
-        # has no guaranted behavior for that, and in fact the current implementation
-        # will keep the first point in the list.
-        datapoints = dict(datapoints).items()
-        state.database.write(metric, datapoints)
-        updateTime = time.time() - t1
+        state.database.create(metric, archiveConfig, xFilesFactor, aggregationMethod)
+        tagMetric(metric)
```

Comment:
Yes, the same way CarbonLink protocol is handled in all the finders. I know that in my implementation I will not want to call Graphite each time I create a metric because it will be too slow.

What can be done is that the default implementation of database.tagMetric() calls Graphite, but can be overloaded to do something else.

Reply:
The problem with doing that is that right now the call `state.database.create` (and `state.database.write`) is blocking, whereas `tagMetric` is designed to be async and use the twisted `HTTPConnectionPool` to avoid blocking the main loop.  On creation of new series it calls `tagMetric` every time (but then moves on and lets the http call complete asynchronously), on updates you can configure a an approximate interval (in number of updates) between calls to `tagMetric` (only really needed to make the tagDB self-healing and ensure consistency).

Reply:
can we add `state.database.tag()` which default implementation is `tagMetric()` ?
This way it's easy to override (for BigGraphite for example), but still does what you want for people who don't override it.